anndata/_core/anndata.py

```diff
@@ -1929,18 +1942,26 @@ def chunk_X(
         else:
             raise ValueError("select should be int or array")
 
-        reverse = None
-        if self.isbacked:
-            # h5py can only slice with a sorted list of unique index values
-            # so random batch with indices [2, 2, 5, 3, 8, 10, 8] will fail
-            # this fixes the problem
-            indices, reverse = np.unique(choice, return_inverse=True)
-            selection = self.X[indices.tolist()]
+        if obsm is not None:
+            selection = self.obsm[obsm][choice]
+        elif layer is not None:
+            selection = self.layers[layer][choice]
         else:
-            selection = self.X[choice]
+            reverse = None
+            if self.isbacked:
+                # h5py can only slice with a sorted list of unique index values
+                # so random batch with indices [2, 2, 5, 3, 8, 10, 8] will fail
+                # this fixes the problem
+                indices, reverse = np.unique(choice, return_inverse=True)
+                selection = self.X[indices.tolist()]
+            else:
+                selection = self.X[choice]
+            selection = selection.toarray() if issparse(selection) else selection
+            selection = selection if reverse is None else selection[reverse]
 
-        selection = selection.toarray() if issparse(selection) else selection
-        return selection if reverse is None else selection[reverse]
+        if obs is not None:
+            selection = (selection, self.obs[obs])
```

Comment:
Would it be like this?

```suggestion
            selection = (selection, self.obs[obs][choice])
```

Reply:
Yep, thank you for catching this.